### ISSUE TITLE
Ability to hide certain fields from the super graph composed by apollo federation

### ISSUE THREAD

**Issue by vamshiaruru-virgodesigns** [strawberry-graphql/strawberry#2044]
Hi all,
In apollo federation 2.0 a new schema directive called `@inaccessible` was introduced. Unfortunately I am not able to find the exact documentation for the directive, but you can see it referred to in https://www.apollographql.com/docs/federation/federation-spec/, and a directive called `Inaccessible` is already present in strawberry.
But to use federation 2.0 directives like inaccessible, you need to set a directive called `link` in each subgraph schema. It is discussed here: https://www.apollographql.com/docs/federation/federation-2/moving-to-federation-2/ . I couldn't find this directive anywhere in strawberry, so I tried implementing it myself, but there are a few issues.

1. The link directive needs a field called `import`, but since import is a python reserved word, it can't be used in python dataclasses. So I am not sure how to proceed here.

2. Just to proceed with testing, I created a directive like this
```python
from dataclasses import field
from typing import List
from strawberry.schema_directive import Location, schema_directive
from strawberry.federation.schema_directives import LinkImport

@schema_directive(locations=[Location.SCHEMA])
class Link:
    url: str
    _import: List[LinkImport] = field(repr="import")
```
and used it like this:
``` python
schema = strawberry.federation.Schema(
    Query,
    Mutation,
    extensions=[
        RelayIdExtension,
        QueryDepthLimiter(max_depth=settings.max_query_depth),
        AddDynamicValidationRules(get_validation_rules),
        OpenTelemetryExtension,
    ],
    directives=[Link(url="https://specs.apollo.dev/federation/v2.0", _import=["@key"])]
)
```
But I get this error when I try that:
```
    schema = strawberry.federation.Schema(
  File "/Users/vamshiaruru/Library/Caches/pypoetry/virtualenvs/ameyam-saleor-app-C_ZSmnnp-py3.9/lib/python3.9/site-packages/strawberry/federation/schema.py", line 27, in __init__
    super().__init__(*args, **kwargs)
  File "/Users/vamshiaruru/Library/Caches/pypoetry/virtualenvs/ameyam-saleor-app-C_ZSmnnp-py3.9/lib/python3.9/site-packages/strawberry/schema/schema.py", line 86, in __init__
    graphql_directives = tuple(
  File "/Users/vamshiaruru/Library/Caches/pypoetry/virtualenvs/ameyam-saleor-app-C_ZSmnnp-py3.9/lib/python3.9/site-packages/strawberry/schema/schema.py", line 87, in <genexpr>
    self.schema_converter.from_directive(directive) for directive in directives
  File "/Users/vamshiaruru/Library/Caches/pypoetry/virtualenvs/ameyam-saleor-app-C_ZSmnnp-py3.9/lib/python3.9/site-packages/strawberry/schema/schema_converter.py", line 142, in from_directive
    for argument in directive.arguments:
AttributeError: 'Link' object has no attribute 'arguments'
```
I dug a little bit into it, and it looks like the schema directives are expected to use `@strawberry.directive.directive` instead of `@schema_directive`? I am not sure what to do here either.

## Feature Request Type

- [ ] Core functionality
- [ ] Alteration (enhancement/optimization) of existing feature(s)
- [x] New behavior

## Description

So in short, a way to use `@link` directive and opt into apollo federation 2 would be great. 
Thanks!

**Comment by patrick91** [comment]
@vamshiaruru-virgodesigns can you try with the following:

```python
@strawberry.type
class Review:
    id: strawberry.ID
    something: str | None = strawberry.federation.field(description="A field that is not used in the schema", inaccessible=True)
```

from my quick test this seems to work already, I'll double check the opt-in to federation 2 directive later, but hopefully using `strawberry.federation.field` should resolve your issue

**Comment by vamshiaruru-virgodesigns** [comment]
Thanks @patrick91 , how would I be able to hide regular mutations/queries using this? Can I use inaccessible keyword in non federation fields as well?

**Comment by patrick91** [comment]
> Thanks @patrick91 , how would I be able to hide regular mutations/queries using this? Can I use inaccessible keyword in non federation fields as well?

`strawberry.federation.field` is a wrapper on top of `strawberry.field` with some additional keywords to make it easier to use federation features, same goes for `strawberry.federation.type` :)

so you can swap `strawberry.field` with `strawberry.federation.field`

I think we are missing `strawberry.federation.mutation`, but I'll fix that in the PR above, there are some other things we are missing 😊

**Comment by vamshiaruru-virgodesigns** [comment]
Thanks!

**Comment by vamshiaruru-virgodesigns** [comment]
Hi @patrick91 , can you tell me how to extend `mutation` to add the same feature? I can add it locally until a new version of strawberry with that feature comes out :)

**Comment by patrick91** [comment]
@vamshiaruru-virgodesigns can you try with this version? `strawberry-graphql==0.124.0.dev.1659561253`

I think I'll do a release today or at worst tomorrow (I just need to finish the docs and fix a typing issue)

**Comment by vamshiaruru-virgodesigns** [comment]
Thanks for the quick response @patrick91 , I'm getting the following error when using this new version (which was not happening in the previous version)
```
  File "/Users/vamshiaruru/Library/Caches/pypoetry/virtualenvs/ameyam-saleor-app-C_ZSmnnp-py3.9/lib/python3.9/site-packages/strawberry/field.py", line 413, in field
    return field_(resolver)
  File "/Users/vamshiaruru/Library/Caches/pypoetry/virtualenvs/ameyam-saleor-app-C_ZSmnnp-py3.9/lib/python3.9/site-packages/strawberry/field.py", line 129, in __call__
    for argument in resolver.arguments:
  File "/opt/homebrew/Cellar/python@3.9/3.9.13_1/Frameworks/Python.framework/Versions/3.9/lib/python3.9/functools.py", line 993, in __get__
    val = self.func(instance)
  File "/Users/vamshiaruru/Library/Caches/pypoetry/virtualenvs/ameyam-saleor-app-C_ZSmnnp-py3.9/lib/python3.9/site-packages/strawberry/types/fields/resolver.py", line 204, in arguments
    for param in user_parameters:
  File "/Users/vamshiaruru/Library/Caches/pypoetry/virtualenvs/ameyam-saleor-app-C_ZSmnnp-py3.9/lib/python3.9/site-packages/strawberry/types/fields/resolver.py", line 203, in <genexpr>
    user_parameters = (p for p in parameters if p not in reserved_parameters)
  File "/opt/homebrew/Cellar/python@3.9/3.9.13_1/Frameworks/Python.framework/Versions/3.9/lib/python3.9/inspect.py", line 2608, in __hash__
    return hash((self.name, self.kind, self.annotation, self.default))
TypeError: unhashable type: 'dict'
```
I have debugged a little, and it happens when I use an empty dictionary as default parameter for any of function params. (I know we are not supposed to use mutable values as default params, but these params were read only).

**Comment by vamshiaruru-virgodesigns** [comment]
That leads to a new error:
```
GraphQL request:1:49
1 | query looks__app__0($channel:String!$first:Int){looks(channel:$channel first:$fi
  |                                                 ^
  | rst){edges{node{id name channelListings{channel{id}}products(first:11){edges{nod
Traceback (most recent call last):
  File "/Users/vamshiaruru/Library/Caches/pypoetry/virtualenvs/ameyam-saleor-app-C_ZSmnnp-py3.9/lib/python3.9/site-packages/graphql/execution/execute.py", line 528, in await_result
    return_type, field_nodes, info, path, await result
  File "/Users/vamshiaruru/Library/Caches/pypoetry/virtualenvs/ameyam-saleor-app-C_ZSmnnp-py3.9/lib/python3.9/site-packages/strawberry/extensions/tracing/opentelemetry.py", line 128, in resolve
    result = await result
  File "/Users/vamshiaruru/Library/Caches/pypoetry/virtualenvs/ameyam-saleor-app-C_ZSmnnp-py3.9/lib/python3.9/site-packages/strawberry/schema/schema_converter.py", line 500, in _async_resolver
    return await await_maybe(_get_result(_source, strawberry_info, **kwargs))
  File "/Users/vamshiaruru/Library/Caches/pypoetry/virtualenvs/ameyam-saleor-app-C_ZSmnnp-py3.9/lib/python3.9/site-packages/strawberry/schema/schema_converter.py", line 482, in _get_result
    field_args, field_kwargs = _get_arguments(
  File "/Users/vamshiaruru/Library/Caches/pypoetry/virtualenvs/ameyam-saleor-app-C_ZSmnnp-py3.9/lib/python3.9/site-packages/strawberry/schema/schema_converter.py", line 419, in _get_arguments
    kwargs = convert_arguments(
  File "/Users/vamshiaruru/Library/Caches/pypoetry/virtualenvs/ameyam-saleor-app-C_ZSmnnp-py3.9/lib/python3.9/site-packages/strawberry/arguments.py", line 195, in convert_arguments
    kwargs[argument.python_name] = convert_argument(
  File "/Users/vamshiaruru/Library/Caches/pypoetry/virtualenvs/ameyam-saleor-app-C_ZSmnnp-py3.9/lib/python3.9/site-packages/strawberry/arguments.py", line 160, in convert_argument
    if graphql_name in value:
TypeError: argument of type 'StrawberryField' is not iterable
Stack (most recent call last):
  File "/Users/vamshiaruru/Library/Caches/pypoetry/virtualenvs/ameyam-saleor-app-C_ZSmnnp-py3.9/lib/python3.9/site-packages/strawberry/fastapi/router.py", line 290, in execute
    return await self.schema.execute(
  File "/Users/vamshiaruru/Library/Caches/pypoetry/virtualenvs/ameyam-saleor-app-C_ZSmnnp-py3.9/lib/python3.9/site-packages/strawberry/schema/schema.py", line 210, in execute
    self.process_errors(result.errors, execution_context=execution_context)
```

This is how I am using it
```
    @strawberry.field
    @inject_session
    async def looks(
        self,
        info: Info,
        channel: str,
        session: AsyncSession,
        before: Optional[str] = None,
        after: Optional[str] = None,
        first: Optional[int] = None,
        last: Optional[int] = None,
        sortBy: LooksSorter = strawberry.field(default_factory=dict),
        filter: LooksFilter = strawberry.field(default_factory=dict),
    ) -> Connection[Look]:
```

Is there a better way of doing it? I want the Sorter/Filter field to be empty by default (in which case no filters will be applied). When I was passing `{}` as the default value, I imagine strawberry was internally doing something like `LooksSorter(**{})` and that used to work.

**Comment by patrick91** [comment]
@vamshiaruru-virgodesigns strawberry.field doesn't work on arguments, though using `{}` seems to work for me here: https://play.strawberry.rocks/?gist=5d30a0078aa6cd1a45c667bda214715e

are you able to share a minimal reproduction? also feel free to message me on discord too

**Comment by vamshiaruru-virgodesigns** [comment]
Thanks Patrick, I'll ping you on discord. But just to help anyone who might stumble onto this, the unhashable type was my issue, not strawberry. I was modifying function params in a decorators, and that was messing things up :)
